src/marqo/core/index_management/index_management.py

```diff
@@ -75,35 +47,66 @@ def __init__(self, vespa_client: VespaClient, zookeeper_client: Optional[Zookeep
+        """
+        Validates index settings using the IndexSettings model.
+
+        Args:
+            index_name (str): The name of the index to validate settings for.
+            settings_dict (dict): A dictionary of settings to validate.
+
+        Raises:
+            ValidationError: If the settings are invalid in the context of the IndexSettings model.
+            api_exceptions.InvalidArgError: If several settings are invalid in the context of the Marqo API.
+            Check the errors in to_marqo_index_request method of IndexSettings model for more details.
+
+        Returns:
+            None: If the validation is successful, nothing is returned,
+            else InvalidArgumentError is raised.
+        """
+        index_settings = IndexSettings(**settings_dict)
+        index_settings.to_marqo_index_request(index_name)
 
     def bootstrap_vespa(self) -> bool:
         """
         Add Marqo configuration to Vespa application package if an existing Marqo configuration is not detected.
 
         Returns:
-            True if Vespa was configured, False if it was already configured
+            True if Vespa was bootstrapped, False if it was already up-to-date
         """
-        app = self.vespa_client.download_application()
-        configured = self._marqo_config_exists(app)
+        with self._vespa_deployment_lock():
+            application = self._get_application_context_manager_for_bootstrapping_and_rollback(check_configured=False)
+            with application as app:
+
+                marqo_version = version.get_version()
+                has_marqo_settings_schema = app.has_schema(self._MARQO_SETTINGS_SCHEMA_NAME)
+                marqo_config_doc = self._get_marqo_config() if has_marqo_settings_schema else None
+
+                if not app.need_bootstrapping(marqo_version, marqo_config_doc):
+                    application.gen.send(False)  # tell context manager to skip deployment
+                    return False
 
-        if not configured:
-            self._add_marqo_config(app)
-            self.vespa_client.deploy_application(app)
-            self.vespa_client.wait_for_application_convergence()
-            self._save_marqo_version(version.get_version())
-            return True
+                existing_indexes = self._get_existing_indexes() if has_marqo_settings_schema else ()
+                app.bootstrap(marqo_version, existing_indexes)
+                return True
 
-        return False
+    def rollback_vespa(self) -> None:
+        with self._vespa_deployment_lock():
+            application = self._get_application_context_manager_for_bootstrapping_and_rollback()
+            with application as app:
+                try:
+                    app.rollback(version.get_version())
+                except ApplicationRollbackError as e:
+                    logger.error(e.message)
+                    application.gen.send(False)  # tell context manager to skip deployment
```

Comment:
What happens if this call fails or we fail before this call? Also does the context manager not suffice?

Reply:
This call tells context manager that we need to skip the deployment due to some failures. You made a good point that this call could fail. But only in two circumstances: 
- the context manager yields all the values already, and raises StopIteration error, which should not happen here since we only use the yielded value once. 
- any exception type when handling the sent value, in this case, it's just a bool value, it should be fine.